### PR TITLE
[HGCal/FW] Add support for HGCal Geom in FW

### PR DESCRIPTION
**Pull request by Mourtz**

### CODE REVIEW

**Comment by Dr15Jones** [comment]
Please remove the `.gitignore` file from this pull request.

**Comment by alja** [comment]
+1

**Comment by alja** [comment]
@mourtz can you add a screenshot of graphical view in the PR?

**Comment by Mourtz** [comment]
@alja 
Here's the Geometry
geom

File: Fireworks/Core/src/FW3DViewGeometry.cc [hunk 1]
```diff
@@ -415,4 +416,90 @@ FW3DViewGeometry::showTrackerEndcap( bool showTrackerEndcap )
    }
 }
 
+//______________________________________________________________________________
+void 
+FW3DViewGeometry::showHGCalEE(bool showHGCalEE){
+   if( showHGCalEE && !m_HGCalEEElements ){
+      m_HGCalEEElements = new TEveElementList("HGCalEE");
+      std::vector<unsigned int> ids = m_geom->getMatchedIds( FWGeometry::HGCalEE );
+      for( std::vector<unsigned int>::const_iterator id = ids.begin();
```
Comment (ianna) [inline Fireworks/Core/src/FW3DViewGeometry.cc]: @Mourtz - please, use a range-based for loop:
```
for( const auto& id : m_geom->getMatchedIds( FWGeometry::HGCalEE ))
```

File: Fireworks/Core/src/FW3DViewGeometry.cc [hunk 2]
```diff
@@ -415,4 +416,90 @@ FW3DViewGeometry::showTrackerEndcap( bool showTrackerEndcap )
...
+      gEve->Redraw3D();
+   }
+}
  
+void 
+FW3DViewGeometry::showHGCalHSi(bool showHGCalHSi){
+  if( showHGCalHSi && !m_HGCalHSiElements ){
+    m_HGCalHSiElements = new TEveElementList("HGCalHSi");
+    std::vector<unsigned int> ids = m_geom->getMatchedIds( FWGeometry::HGCalHSi );
+    for( std::vector<unsigned int>::const_iterator id = ids.begin();
```
Comment (ianna) [inline Fireworks/Core/src/FW3DViewGeometry.cc]: @Mourtz - please, use a range-based for loop

File: Fireworks/Core/src/FW3DViewGeometry.cc [hunk 3]
```diff
@@ -415,4 +416,90 @@ FW3DViewGeometry::showTrackerEndcap( bool showTrackerEndcap )
...
+    gEve->Redraw3D();
+  }
+}
+
+void FW3DViewGeometry::showHGCalHSc(bool showHGCalHSc){
+   if( showHGCalHSc && ! m_HGCalHScElements )
+   {
+      m_HGCalHScElements = new TEveElementList( "HGCalHSc" );
+      std::vector<unsigned int> ids = m_geom->getMatchedIds( FWGeometry::HGCalHSc );
+      for( std::vector<unsigned int>::const_iterator id = ids.begin();
```
Comment (ianna) [inline Fireworks/Core/src/FW3DViewGeometry.cc]: @Mourtz - please, use a range-based for loop

File: Fireworks/Core/src/FWGeometry.cc [hunk 4]
```diff
@@ -225,6 +225,34 @@ FWGeometry::getMatchedIds( Detector det, SubDetector subdet ) const
    return ids;
 }
 
+std::vector<unsigned int>
+FWGeometry::getMatchedIds( Detector det ) const
+{
+   std::vector<unsigned int> ids;
+   for( IdToInfoItr it = m_idToInfo.begin(), itEnd = m_idToInfo.end();
```
Comment (ianna) [inline Fireworks/Core/src/FWGeometry.cc]: @Mourtz - please, use a range-based for loop

File: Fireworks/Core/src/FWGeometry.cc [hunk 5]
```diff
@@ -225,6 +225,34 @@ FWGeometry::getMatchedIds( Detector det, SubDetector subdet ) const
...
+   {
+      if( (( it->id >> kDetOffset ) & 0xF) != det ) continue;
+      // select only the 1st layer
+      // if(((it->id>>17)&0x1F) != 12) continue;
+
+      // select only the first cell of each wafer
+      if(det != HGCalHSc){
+        bool flag(false);
+
+        for(std::vector<unsigned int>::iterator id_it = ids.begin(); id_it != ids.end(); ++id_it) {
```
Comment (ianna) [inline Fireworks/Core/src/FWGeometry.cc]: @Mourtz - please, use a range-based for loop or should it be ```find_if```?

File: Fireworks/Core/src/FWGeometry.cc [hunk 6]
```diff
@@ -293,6 +321,111 @@ FWGeometry::getEveShape( unsigned int id  ) const
...
+TEveGeoShape*
+FWGeometry::getHGCSiliconEveShape( unsigned int id  ) const
+{
+#if 0 
+   const unsigned int type = (id>>26)&0x3;
+   // select the middle cell of each waifer
+   id &= ~0x3FF;
+   id |= (type == 0) ? 0x16B : 0xE7;
+#else
+   float sideToSideWaferSize = 16.7441;
```
Comment (ianna) [inline Fireworks/Core/src/FWGeometry.cc]:
```
16.7441f
```
Please use the literal here and also later in the code

File: Fireworks/Core/src/FWGeometry.cc [hunk 7]
```diff
@@ -293,6 +321,111 @@ FWGeometry::getEveShape( unsigned int id  ) const
...
+   float dx = sideToSideWaferSize/2;
```
Comment (ianna) [inline Fireworks/Core/src/FWGeometry.cc]:
```
0.5f*sideToSideWaferSize
```

File: Fireworks/SimData/plugins/FWCaloParticleProxyBuilder.cc [hunk 8]
```diff
@@ -12,77 +12,96 @@
...
-   boxset->UseSingleColor();
-   boxset->SetPickable(true);
-   for (const auto & c : iData.simClusters())
+FWCaloParticleProxyBuilder::build( const FWEventItem* iItem, TEveElementList* product, const FWViewContext* ){
+   const CaloParticleCollection* collection = nullptr;
+   iItem->get( collection );
+
+   if( collection == nullptr ) return;
+
+   for( std::vector<CaloParticle>::const_iterator iData = collection->begin(), end = collection->end(); iData != end; ++iData )
```
Comment (ianna) [inline Fireworks/SimData/plugins/FWCaloParticleProxyBuilder.cc]:
```
for( const auto& iData : collection )
```

File: Fireworks/SimData/plugins/FWCaloParticleProxyBuilder.cc [hunk 9]
```diff
@@ -12,77 +12,96 @@
...
+      TEveBoxSet* boxset = new TEveBoxSet();
+      boxset->Reset(TEveBoxSet::kBT_FreeBox, true, 64);
+      boxset->UseSingleColor();
+      boxset->SetPickable(true);
+
+      for (const auto & c : iData->simClusters())
+      {
+         auto clusterDetIds = (*c).hits_and_fractions();
+
+         for( auto it = clusterDetIds.begin(), itEnd = clusterDetIds.end();
```
Comment (ianna) [inline Fireworks/SimData/plugins/FWCaloParticleProxyBuilder.cc]:
```
for( auto it : clusterDetIds )
```

File: Fireworks/SimData/plugins/FWCaloParticleProxyBuilder.cc [hunk 10]
```diff
@@ -12,77 +12,96 @@
...
+      }
+      boxset->RefitPlex();
+      setupAddElement(boxset, product);
    }
-   boxset->RefitPlex();
-   setupAddElement(boxset, &oItemHolder);
 }
 
-REGISTER_FWPROXYBUILDER( FWCaloParticleProxyBuilder, CaloParticle, "CaloParticles", FWViewType::kAll3DBits | FWViewType::kAllRPZBits );
+REGISTER_FWPROXYBUILDER( FWCaloParticleProxyBuilder, CaloParticleCollection, "CaloFTW", FWViewType::kAll3DBits | FWViewType::kAllRPZBits );
```
Comment (ianna) [inline Fireworks/SimData/plugins/FWCaloParticleProxyBuilder.cc]: please, remove the special character at the end

File: Geometry/HGCalGeometry/src/HGCalGeometry.cc [hunk 11]
```diff
@@ -315,38 +317,93 @@ HGCalGeometry::CornersVec HGCalGeometry::getCorners(const DetId& id) const {
...
   
   unsigned int ncorner = FlatTrd::ncorner_;
   HGCalGeometry::CornersVec co (ncorner, GlobalPoint(0,0,0));
-  unsigned int cellIndex =  indexFor(id);
+  unsigned int cellIndex =  indexFor(detid);
   if ((cellIndex <  m_cellVec.size() && m_det != DetId::HGCalHSc) ||
       (cellIndex <  m_cellVec2.size() && m_det == DetId::HGCalHSc)) {
-    HGCalTopology::DecodedDetId id_ = m_topology.decode(id);
+    HGCalTopology::DecodedDetId id = m_topology.decode(detid);
     std::pair<float,float> xy;
```
Comment (ianna) [inline Geometry/HGCalGeometry/src/HGCalGeometry.cc]: @Mourtz - this is exact code duplication from line 281. Please, implement it as a separate private member function.

File: Geometry/HGCalGeometry/src/HGCalGeometry.cc [hunk 12]
```diff
@@ -315,38 +317,93 @@ HGCalGeometry::CornersVec HGCalGeometry::getCorners(const DetId& id) const {
...
+HGCalGeometry::CornersVec HGCalGeometry::getNewCorners(const DetId& detid) const {
+
+  unsigned int ncorner = (m_det == DetId::HGCalHSc) ? 5 : 7;
+  HGCalGeometry::CornersVec co (ncorner, GlobalPoint(0,0,0));
+  unsigned int cellIndex =  indexFor(detid);
+  if ((cellIndex <  m_cellVec.size() && m_det != DetId::HGCalHSc) ||
+      (cellIndex <  m_cellVec2.size() && m_det == DetId::HGCalHSc)) {
+    HGCalTopology::DecodedDetId id = m_topology.decode(detid);
+    int signz = (id.zSide >= 0) ? -1 : 1;
+    std::pair<float,float> xy;
```
Comment (ianna) [inline Geometry/HGCalGeometry/src/HGCalGeometry.cc]: @Mourtz - is it the same code as above?

**Review by ianna — CHANGES_REQUESTED** [review]

**Comment by kpedro88** [comment]
please test

**Comment by kpedro88** [comment]
+upgrade
The image in https://github.com/cms-sw/cmssw/pull/25135#issuecomment-436680639 looks very nice

File: Geometry/HGCalGeometry/test/HGCalTestNeighbor.cc [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+	    << geom->cellElement() << std::endl;
+  int layers[] = {9, 15, 22};
+  int zsides[] = {1, -1};
+  int iphis[]  = {1, 51, 101, 151, 201};
+  int ietas[]  = {11, 20, 29};
+  for (int zside : zsides) {
+    for (int layer : layers) {
+      int type = geom->topology().dddConstants().getTypeTrap(layer);
+      for (int ieta : ietas) {
+	for (int iphi : iphis) {
```
Comment (ianna) [inline Geometry/HGCalGeometry/test/HGCalTestNeighbor.cc]: @Mourtz - an indentation is missing

File: Geometry/HGCalGeometry/src/HGCalGeometry.cc (hunk 11 above)
Comment (ianna) [reply Geometry/HGCalGeometry/src/HGCalGeometry.cc]: @Mourtz - I do not see where it was resolved.

**Review by ianna — CHANGES_REQUESTED** [review]

File: Geometry/HGCalGeometry/test/HGCalTestNeighbor.cc (hunk 13 above)
Comment (Mourtz) [reply Geometry/HGCalGeometry/test/HGCalTestNeighbor.cc]: Merged this code form #25027.

**Comment by Mourtz** [comment]
@ianna 
The files that you are suggesting me to modify were merged from #25027.

**Comment by Mourtz** [comment]
@ianna 
It shouldn't show any changes in those files....
Do you want me to get them out the PR?

**Comment by Mourtz** [comment]
is there any problem with merging the PR?

**Comment by ianna** [comment]
+1

**Comment by fabiocos** [comment]
@alja could you please check and sign it in case?

**Comment by alja** [comment]
+1

**Comment by alja** [comment]
+1

**Comment by Mourtz** [comment]
changes of this PR are included in #25284. Keeping only the last PR.

**Pull request closed without merging**